### PR TITLE
Make sure all errors messages are correctly encoded

### LINKED ISSUE #587: Make sure all errors messages are correctly encoded
See https://github.com/openfisca/openfisca-core/issues/580

If a user sends a non ASCII character to OpenFisca, there may be an encoding error raised by OpenFisca.

Some `u` is probably missing somewhere.

### PR DESCRIPTION
**Pull request by sandcha**
Fixes #587 
Connected to #580

#### Bug fix

- Fix API response encoding from `ascii` to `utf-8` (to manage non `ascii` characters).
  - Improve user message by displaying `UnicodeDecodeError` information in json error response

### CODE REVIEW

File: openfisca_core/entities.py [hunk 1]
```diff
@@ -71,7 +71,7 @@ def init_variable_values(self, entity_object, entity_id):
 
             if not isinstance(variable_values, dict):
                 raise SituationParsingError(path_in_json,
-                    'Invalid type: must be of type object. Input variables must be set for specific periods. For instance: {"salary": {"2017-01": 2000, "2017-02": 2500}}')
+                    u'Invalid type: must be of type object. Input variables must be set for specific periods. For instance: {"salary": {"2017-01": 2000, "2017-02": 2500}}')
```
Comment (Anna-Livia) [inline openfisca_core/entities.py]: When tested with curl, got : 
```
 For instance: {\"salary\": {\"2017-01\": 2000, \"2017-02\": 2500}}"
```
I was expecting : 
```
 For instance: {"salary": {"2017-01": 2000, "2017-02": 2500}}"
```

File: openfisca_core/entities.py [hunk 2]
```diff
@@ -89,15 +89,14 @@ def init_variable_values(self, entity_object, entity_id):
                             value = holder.variable.possible_values[value]
                         except KeyError:
                             raise SituationParsingError(path_in_json,
-                                "'{}' is not a valid value for '{}'. Possible values are ['{}'].".format(
-                                    value, variable_name, "', '".join(holder.variable.possible_values.list)
-                                    ).encode('utf-8')
+                                u"'{}' is not a valid value for '{}'. Possible values are ['{}'].".format(
```
Comment (Anna-Livia) [inline openfisca_core/entities.py]: Works great when Value has accents. Woudl it be possible to add a accentfull enum to country template to test if possible values works also ?

File: openfisca_core/entities.py [hunk 3]
```diff
@@ -89,15 +89,14 @@ def init_variable_values(self, entity_object, entity_id):
...
+                                    value, variable_name, "', '".join(holder.variable.possible_values.list))
                                 )
                     try:
                         array[entity_index] = value
                     except (ValueError, TypeError) as e:
                         raise SituationParsingError(path_in_json,
-                    'Invalid type: must be of type {}.'.format(holder.variable.json_type))
+                            u'Invalid type: must be of type {}.'.format(holder.variable.json_type))
```
Comment (Anna-Livia) [inline openfisca_core/entities.py]: I don't see how to test for this 🤔

File: openfisca_core/entities.py [hunk 4]
```diff
@@ -115,7 +114,7 @@ def finalize_variables_init(self):
                     # It is only raised when we consume the buffer. We thus don't know which exact key caused the error.
                     # We do a basic research to find the culprit path
                     culprit_path = next(
-                        dpath.search(self.entities_json, "*/{}/{}".format(e.variable_name, str(e.period)), yielded = True),
+                        dpath.search(self.entities_json, u"*/{}/{}".format(e.variable_name, str(e.period)), yielded = True),
```
Comment (Anna-Livia) [inline openfisca_core/entities.py]: I don't see how to test for this 🤔

File: openfisca_core/entities.py [hunk 5]
```diff
@@ -145,7 +144,7 @@ def clone(self, new_simulation):
     def __getattr__(self, attribute):
         projector = get_projector_from_shortcut(self, attribute)
         if not projector:
-            raise Exception("Entity {} has no attribute {}".format(self.key, attribute))
+            raise Exception(u"Entity {} has no attribute {}".format(self.key, attribute))
```
Comment (Anna-Livia) [inline openfisca_core/entities.py]: I don't see how to test for this 🤔

File: openfisca_core/entities.py [hunk 6]
```diff
@@ -319,12 +318,12 @@ def init_members(self, roles_json, entity_id):
                 check_type(person_id, basestring, [self.plural, entity_id, role_id, str(index)])
                 if person_id not in self.simulation.persons.ids:
                     raise SituationParsingError([self.plural, entity_id, role_id],
-                        "Unexpected value: {0}. {0} has been declared in {1} {2}, but has not been declared in {3}.".format(
+                        u"Unexpected value: {0}. {0} has been declared in {1} {2}, but has not been declared in {3}.".format(
```
Comment (Anna-Livia) [inline openfisca_core/entities.py]: Tested, works well :)

File: openfisca_core/entities.py [hunk 7]
```diff
@@ -305,7 +304,7 @@ def init_from_json(self, entities_json):
         if self.persons_to_allocate:
             unallocated_person = self.persons_to_allocate.pop()
             raise SituationParsingError([self.plural],
-                '{0} has been declared in {1}, but is not a member of any {2}. All {1} must be allocated to a {2}.'.format(
+                u'{0} has been declared in {1}, but is not a member of any {2}. All {1} must be allocated to a {2}.'.format(
```
Comment (Anna-Livia) [inline openfisca_core/entities.py]: Works well :)

File: openfisca_core/simulations.py [hunk 8]
```diff
@@ -254,7 +254,7 @@ def check_type(input, type, path = []):
 
     if not isinstance(input, type):
         raise SituationParsingError(path,
-            "Invalid type: must be of type '{}'.".format(json_type_map[type]))
+            u"Invalid type: must be of type '{}'.".format(json_type_map[type]))
```
Comment (Anna-Livia) [inline openfisca_core/simulations.py]: 👍

**Comment by sandcha** [comment]
On enums with accents:
Before merging enums pr #589, enums with accents work. This is conditioned to the use of unicode on key names. Str keys with accent raise `KeyError` on `enumerations.py`:16. But this file disappears in #589.

File: openfisca_core/entities.py (hunk 1 above)
Comment (sandcha) [reply openfisca_core/entities.py]: `For instance: {\"salary\": {\"2017-01\": 2000, \"2017-02\": 2500}}` is correct as it is included in a JSON response like: 
```
{
  "persons": {
    "bill": {
      "salary": "Invalid type: must be of type object. Input variables must be set for specific periods. For instance: {\"salary\": {\"2017-01\": 2000, \"2017-02\": 2500}}"
    }
  }
}
```

**Comment by sandcha** [comment]
Definition of done:
- [] When a non ASCII encoding is sent to the API, print a 500 error message with the incriminated character.
- [] Add unit tests on encoding in API requests

**Comment by sandcha** [comment]
Before, on `UnicodeDecodeError `:
```
{
  "error": "Internal server error: 'ascii' codec can't encode character u'\\u2018' in position 1: ordinal not in range(128)"
}
```

After:
```
{
  "error": "Internal server error: 'O‘Ryan' is not a valid ASCII value."
}
```

**Pull request merged**